// template-helper/lib/helper.js
/**
 * Created by patrickliu on 15/7/1.
 */

var dateTool = require('node-date-tool'),
    querystring = require('querystring'),
    url = require('url'),
    safeJSONStringify = require('safe-echo-json'),
    arabicNumberToRMB = require('node-arabic-number-to-rmb'),
    formatNumber = require('make-number-more-readable'),
    arithmetic = require('./arithmetic');


var helperArr = {

    // dateTool
    dateTool: dateTool,

    // safeJSONStringify
    safeJSONStringify: safeJSONStringify,

    // pass Date Object
    Date: Date,

    // pass parseFloat
    parseFloat: parseFloat,

    // parse int
    parseInt: parseInt,

    // tofixed
    toFixed: toFixed,

    // add arabicnumber to rmb format
    arabicNumberToRMB: arabicNumberToRMB,

    // format number 1000000 -> 1,000,000
    formatNumber: formatNumber,

    // expose the 4 methods below
    decodeURIComponent: decodeURIComponent,
    encodeURIComponent: encodeURIComponent,
    encodeURI: encodeURI,
    decodeURI: decodeURI,

    // expose node module, url & querystring
    nodeUrl: url,
    nodeQueryString: querystring,

    // expose RegExp
    RegExp: RegExp,

    // expose Math Object
    Math: Math,
    
    // 简易的加减乘除运算
    arithmetic: arithmetic

};

function toFixed(number, fixed) {
    return Number(number).toFixed(fixed);
}

exports = module.exports = helperArr;
